resolver/resolver.go

```diff
@@ -2415,6 +2723,181 @@ func (r *Resolver) resolveMessageCELValues(ctx *context, env *cel.Env, msg *Mess
+		}
+		r.resolveMapIteratorExprCELValues(ctx, mapEnv, expr.Map.Expr)
+		varType := expr.Map.Expr.Type.Clone()
+		varType.Repeated = true
+		expr.Type = varType
+	case expr.Call != nil:
+		if expr.Call.Request != nil {
+			for idx, arg := range expr.Call.Request.Args {
+				if arg.Value == nil {
+					continue
+				}
+				if err := r.resolveCELValue(ctx, env, arg.Value.CEL); err != nil {
+					ctx.addError(
+						ErrWithLocation(
+							err.Error(),
+							source.CallExprRequestByLocation(
+								ctx.fileName(),
+								ctx.messageName(),
+								ctx.defIndex(),
+								idx,
+							),
+						),
+					)
+				}
+			}
+		}
+		expr.Type = NewMessageType(expr.Call.Method.Response, false)
+	case expr.Message != nil:
+		for argIdx, arg := range expr.Message.Args {
+			if arg.Value == nil {
+				continue
+			}
+			if err := r.resolveCELValue(ctx, env, arg.Value.CEL); err != nil {
+				if arg.Value.CEL != nil && arg.Value.Inline {
+					ctx.addError(
+						ErrWithLocation(
+							err.Error(),
+							source.MessageExprArgumentInlineLocation(
+								ctx.fileName(),
+								ctx.messageName(),
+								ctx.defIndex(),
+								argIdx,
+							),
+						),
+					)
+				} else {
+					ctx.addError(
+						ErrWithLocation(
+							err.Error(),
+							source.MessageExprArgumentByLocation(
+								ctx.fileName(),
+								ctx.messageName(),
+								ctx.defIndex(),
+								argIdx,
+							),
+						),
+					)
+				}
+			}
+		}
+		expr.Type = NewMessageType(expr.Message.Message, false)
+	case expr.Validation != nil:
+		if expr.Validation.Error.Rule != nil {
+			e := expr.Validation.Error
+			if err := r.resolveCELValue(ctx, env, e.Rule); err != nil {
+				ctx.addError(
+					ErrWithLocation(
+						err.Error(),
+						source.MessageValidationLocation(ctx.fileName(), ctx.messageName(), ctx.defIndex(), true),
+					),
+				)
+				return
+			}
+			if e.Rule.Out.Type != types.Bool {
+				ctx.addError(
+					ErrWithLocation(
+						"validation rule must always return a boolean value",
+						source.MessageValidationLocation(ctx.fileName(), ctx.messageName(), ctx.defIndex(), true),
+					),
+				)
+			}
+		}
+		for detIdx, detail := range expr.Validation.Error.Details {
+			r.resolveMessageValidationErrorDetailCELValues(ctx, env, ctx.msg, ctx.defIndex(), detIdx, detail)
+		}
+	}
+}
+
+func (r *Resolver) resolveMapIteratorExprCELValues(ctx *context, env *cel.Env, expr *MapIteratorExpr) {
+	switch {
+	case expr.By != nil:
+		if err := r.resolveCELValue(ctx, env, expr.By); err != nil {
+			ctx.addError(
+				ErrWithLocation(
+					err.Error(),
+					source.MapExprByLocation(
+						ctx.fileName(),
+						ctx.messageName(),
+						ctx.defIndex(),
+					),
+				),
+			)
+		}
+		expr.Type = expr.By.Out
+	case expr.Message != nil:
+		for argIdx, arg := range expr.Message.Args {
+			if arg.Value == nil {
+				continue
+			}
+			if err := r.resolveCELValue(ctx, env, arg.Value.CEL); err != nil {
+				if arg.Value.CEL != nil && arg.Value.Inline {
```

Comment:
nit, IIUC if `arg.Value.CEL` is nil, `resolveCELValue` never returns an error, so `arg.Value.CEL != nil` is unnecessary.